test/hotspot/jtreg/serviceability/jvmti/vthread/FollowReferences/VThreadStackRefTest.java

```diff
@@ -0,0 +1,157 @@
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/**
+ * @test
+ * @requires vm.jvmti
+ * @requires vm.continuations
+ * @enablePreview
+ * @run main/othervm/native
+ *      -Djava.util.concurrent.ForkJoinPool.common.parallelism=1
+ *      -agentlib:VThreadStackRefTest
+ *      VThreadStackRefTest
+ */
+import java.util.List;
+import java.util.ArrayList;
+import java.util.stream.Stream;
+import java.util.concurrent.CountDownLatch;
+
+public class VThreadStackRefTest {
+
+    static volatile boolean timeToStop = false;
+    static int i = -1;
+
+    public static void main(String[] args) throws InterruptedException {
+        CountDownLatch dumpedLatch = new CountDownLatch(1);
+        Thread vthreadMounted = Thread.ofVirtual().start(() -> {
+            Object referenced = new VThreadMountedReferenced();
+            System.out.println(referenced.getClass());
+            while (!timeToStop) {
+                if (++i == 10000) {
+                    i = 0;
+                }
+            }
+            await(dumpedLatch);
+            System.out.println(referenced.getClass());
+        });
+        Thread vthreadUnmounted = Thread.ofVirtual().start(() -> {
+            Object referenced = new VThreadUnmountedReferenced();
+            System.out.println(referenced.getClass());
+            await(dumpedLatch);
+            System.out.println(referenced.getClass());
+        });
+        Thread vthreadJNIUnmounted = Thread.ofVirtual().start(() -> {
+            createObjAndCallback(VThreadUnmountedJNIReferenced.class,
+                new Runnable() {
+                    public void run() {
+                        await(dumpedLatch);
+                    }
+                });
+        });
+
+        Thread vthreadEnded = Thread.ofVirtual().start(() -> {
+            Object referenced = new VThreadUnmountedEnded();
+            System.out.println(referenced.getClass());
+        });
+        Thread pthread = Thread.ofPlatform().start(() -> {
+            Object referenced = new PThreadReferenced();
+            System.out.println(referenced.getClass());
+
+            await(dumpedLatch);
+            System.out.println(referenced.getClass());
+        });
+        vthreadEnded.join();
+
+        Thread.sleep(2000); // wait for reference and unmount
```

Comment:
I think what you mean is you need to wait until the threads have made enough progress to create the references, and then you need to wait until they have had a chance to amount due to the await() call. This should be made more clear in the comments.

BTW, you could choose to get JVMTI VIRTUAL_THREAD_UNMOUNT events, and instead block here until you get them all, but doing a sleep is a lot easier.

Reply:
Added comment.
Sleep does the job, I don't think it makes sense to overcomplicate the test